File: exercice4_2.c
#include<stdio.h>
#include<stdlib.h>
#include<sys/wait.h>
#include<unistd.h>
#include<fcntl.h>
#include<string.h>

#define SIZECAR 100
int main()
{
    pid_t pid_fils;
    int tube1[2];
    int tube2[2];
    char caractereW[100];

    //creation du pipe
    if(pipe(tube1)!=0 || pipe(tube2)!=0)
    {
        fprintf(stderr, "Erreur dans pipe\n");
        exit(1);
    }

    pid_fils = fork();

    if (pid_fils == -1)
    {
        fprintf(stderr, "Erreur dans fork\n");
        exit(1);
    }
    if(pid_fils == 0)
    {
        //----------------pipe1_------------------

        printf("saisissez un mot\n");
        scanf("%s",caractereW);

        close(tube1[0]);
        close(tube2[1]);
        
        dup2(tube1[1] ,STDOUT_FILENO);
        dup2(tube2[0],STDIN_FILENO);

        close(tube1[1]);
        close(tube2[0]);
        
        //---------------pipe_2-------------------

        printf("%s\n",caractereW);//envoyer père
        fflush(stdout);


        int resultat;

        scanf("%d",&resultat);

        if(resultat == 1 )
        {
            fprintf(stderr,"le mot existe dans le fichier\n");
        }
        else
        {
            fprintf(stderr,"le mot n'existe pas dans le fichier\n");
        }
        
    }
    else
    {
        //-------------------pipe_1----------------
        close(tube1[1]);//pas ecriture
        close(tube2[0]);//pas lecture

        dup2(tube1[0], STDIN_FILENO);
        dup2(tube2[1],STDOUT_FILENO);

        close(tube1[0]);
        close(tube2[1]);

        
        char caractereR[50];
        scanf("%s",caractereR);//lire mot venant du fils

        char ligne[50];
        int compt=0;

        FILE *mot=fopen("fichier.txt","r");//ouverture du fichier contenant le mot
        if(mot==NULL)
        {
            perror("fopen");
            exit(1);
        }

        while(fscanf(mot,"%s",ligne)!=EOF) //verification mot par mot
        {
            if(strcmp(ligne,caractereR)==0)//comparaison des chaines de caratcère
            {
                compt=1;
                break;
            }
        }

       fclose(mot);

       //---------------pipe_2--------------------
       printf("%d\n",compt);//envoyer au processus fils
       fflush(stdout);

       exit(0);
    }

    return(0);

}

File: exercice4_2.py
import os
import sys

def main():
    perevfils_r, perevfils_w = os.pipe()
    filsvpere_r, filsvpere_w = os.pipe()

    nom_fic = "monFichier"

    fils = os.fork()

    if fils == 0:
        # ===== FILS =====
        mot = input("veuillez entrer le mot à chercher : ")

        os.close(filsvpere_r)
        os.close(perevfils_w)

        os.dup2(filsvpere_w, 1)  # STDOUT
        os.dup2(perevfils_r, 0)  # STDIN

        os.close(filsvpere_w)
        os.close(perevfils_r)

        print(mot, end='', flush=True)

        try:
            x = int(input())
        except:
            x = 0

        if x == 0:
            print("on a pas trouver le mot dans le fichier")
            os._exit(3)
        else:
            print("la recherche a reussi")
            os._exit(3)

    elif fils > 0:
        # ===== PERE =====
        os.close(filsvpere_w)
        os.close(perevfils_r)

        os.dup2(filsvpere_r, 0)  # STDIN
        os.dup2(perevfils_w, 1)  # STDOUT

        os.close(filsvpere_r)
        os.close(perevfils_w)

        try:
            recuperer = input()
        except:
            recuperer = ""

        print(f"recuperer vaut : {recuperer}")

        trouver = 0

        try:
            with open(nom_fic, "r") as ptr:
                for ligne in ptr:
                    mots = ligne.split()
                    for chaine in mots:
                        if chaine == recuperer:
                            trouver = 1
                            break
                    if trouver:
                        break
        except:
            pass

        print(trouver, end='', flush=True)

if __name__ == "__main__":
    main()